lib/Twig/Environment.php

```diff
@@ -326,15 +363,21 @@ public function loadTemplate($name, $index = null)
         }
 
         if (!class_exists($cls, false)) {
-            if (false === $cache = $this->getCacheFilename($name)) {
-                eval('?>'.$this->compileSource($this->getLoader()->getSource($name), $name));
+            if ($this->bcGetCacheFilename) {
+                $key = $this->getCacheFilename($name);
             } else {
-                if (!is_file($cache) || ($this->isAutoReload() && !$this->isTemplateFresh($name, filemtime($cache)))) {
-                    $this->writeCacheFile($cache, $this->compileSource($this->getLoader()->getSource($name), $name));
-                }
+                $key = $this->cache->generateKey($cls, $this->templateClassPrefix);
```

Comment:
Understood. I was not happy to leak such an implementation detail (the class prefix) and #1822 tries to fix your issue and provide a much better UX. Tell me if that works for you.